MAINT Introduce FastEuclideanPairwiseArgKmin

jjerphan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Part of splitting https://github.com/scikit-learn/scikit-learn/pull/21462#issuecomment-993862788

⚠ This targets the `upstream:pairwise-distances-argkmin` feature branch not `upstream:main`.

#### What does this implement/fix? Explain your changes.

This introduces `FastEuclideanPairwiseDistancesArgKmin` a specialization of `PairwiseDistancesArgkmin` for the euclidean and squared euclidean distances which uses the GEMM trick.

#### Any other comments

I will rebase and force-push on update made to https://github.com/scikit-learn/scikit-learn/pull/22064.

thomasjpfan commented:
Marking this as a draft for now until https://github.com/scikit-learn/scikit-learn/pull/22064 is merged.

jjerphan committed:
MAINT Introduce FastEuclideanPairwiseArgKmin

This reverts the main changes made by 09a9527
to make the initialization in __cinit__ instead
of in __init__ because it's easier this way.

If there's a way to maintain the initialization
in __cinit__, let's do it.

jjerphan committed:
fixup! MAINT Introduce FastEuclideanPairwiseArgKmin

jjerphan committed:
Fix imports and typos

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -629,15 +664,16 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
         return pda._finalize_results(return_distance)
 
-    def __cinit__(
+    def __init__(
```
Based on my understanding of `__cinit__`, we need to do something like this:

```python
cdef class PairwiseDistancesReduction:
    def __cinit__(...):
        self.heaps_r_distances_chunks = <DTYPE_t **> malloc(
            sizeof(DTYPE_t *) * self.chunks_n_threads
        )
        ...
```

and in the subclass `FastEuclideanPairwiseDistancesArgKmin`:

```python
cdef class FastEuclideanPairwiseDistancesArgKmin:
    def __cinit__(...):
        # Do not need to call `super()` here. Cython handles it for you
        self.dist_middle_terms_chunks = <DTYPE_t **> malloc(
             sizeof(DTYPE_t *) * self.chunks_n_threads
         )
```

Based on [cython docs](https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init):

> If your extension type has a base type, any existing __cinit__() methods in the base type hierarchy are automatically called before your __cinit__() method.

jjerphan replied:
Yes, it's just that `__init__` has to have the same signature and arguments than `__cinit__`.

It might be possible via using args and kwargs, but might also make the signature more complex. I am currently thinking about it.

jjerphan replied:
03b6c79 resolves this.

thomasjpfan replied:
Seeing the diff to make `__cinit__` work, I am -1 on the idea because of how much more complexity it adds.

Do you think there are the consequences for using `__init__` for `malloc`?

jjerphan replied:
> Seeing the diff to make __cinit__ work, I am -1 on the idea because of how much more complexity it adds.

So am I.

> Do you think there are any consequences for using __init__ for malloc?

`__cinit__` is always called. There's no such guarantee for `__init__`.
In some cases (e.g. for de-serialization or when creating an instance via `__new__`), `__init__` won't be called.

If those interfaces were to be public, such problem would appear (because data-structures would not be allocated on deserialization in the user code).
 
In our case, I think it is OK-ish, because:
 - they are allocated in Python via `__init__` in a private implementation. 
 - those aren't attributes of estimators hence we do not have the problem with deserialization.
 
I would be in favor of reverting everything to use `__init__` and add a comment motivating this choice.

What do you think?

jjerphan replied:
I reverted those changes to solely use `__init__` in the last commits.

Let me know if this works for you.

thomasjpfan replied:
> Let me know if this works for you.

It works for me.

jjerphan committed:
TST Adapt test_pairwise_distances_argkmin

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
MAINT Reorder initilizations to move allocations in __cinit__

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Revert "MAINT Reorder initilizations to move allocations in __cinit__"

This reverts commit 03b6c79e04c8d0a23a23493a6426907dd982a9ea.

jjerphan committed:
Simply plug FastEuclideanPairwiseDistancesArgKmin in

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -837,3 +880,204 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
+    better running time when the alternative is IO bound, but it can suffer
+    from numerical instability caused by catastrophic cancellation potentially
+    introduced by the subtraction in the arithmetic expression above.
+
+    PairwiseDistancesArgKmin with EuclideanDistance must be used when higher
+    numerical precision is needed.
```
```suggestion
```

jjerphan committed:
Remove old remark

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -611,13 +644,26 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
+                strategy=strategy,
+                metric_kwargs=metric_kwargs,
+            )
+        else:  # Fall back on the default
```
```suggestion
        else:
             # Fall back on a generic implementation that handles most scipy
             # metrics by computing the distances between 2 vectors at a time.
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction.pyx:
```diff
@@ -611,13 +644,26 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
+            and not issparse(X)
+            and not issparse(Y)
+        ):
+            use_squared_distances = metric == "sqeuclidean"
```
```suggestion
            # Specialized implementation with improved arithmetic intensity
            # and vector instructions (SIMD) by processing several vectors
            # at time to leverage a call to the BLAS GEMM routine as explained
            # in more details in the docstring.
            use_squared_distances = metric == "sqeuclidean"
```

ogrisel reviewed (approved):
LGTM!

jjerphan committed:
Add indications for the dispatch of the implementations

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented:
I think we can ignore the code coverage report. It seems inconsistent and it's probably because this sub-PR does not target `main`.

ogrisel commented:
@thomasjpfan do you have other things in mind w.r.t. this PR?

thomasjpfan reviewed (approved):
Minor comments, otherwise LGTM

jjerphan committed:
Simplify and do not use indirection

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

ogrisel commented:
Merged!
